R/liger.R

```diff
@@ -2030,29 +2030,35 @@ imputeKNN <- function(object, reference, queries = NULL, knn_k = 50, weight = FA
         paste("  ", reference, "\n"),
         "Query datasets:\n",
         paste("  ", as.character(queries), "\n")
-        )
-    }
+      )
     }
+  }
   
   reference_cells <- rownames(object@scale.data[[reference]]) # cells by genes
   for (query in queries) {
     query_cells <- rownames(object@scale.data[[query]])
-    # find nearest neighbors for query cell in normed ref datasets
-    nn.k <- get.knnx(object@H.norm[reference_cells, ], object@H.norm[query_cells, ], k = knn_k, algorithm = "CR")
-    imputed_vals <- sapply(1:nrow(nn.k$nn.index), function(n) { # for each cell in the target dataset:
-      weights <- nn.k$nn.dist[n, ]
-      weights <- as.matrix(exp(-weights) / sum(exp(-weights)))
-      imp <- object@raw.data[[reference]][, nn.k$nn.index[n, ]] # genes by cells, genes are from reference dataset
-      if (weight) {
-        imp <- as.matrix(imp %*% weights) # (genes by k) multiply by the weight matrix (k by 1)
+    
+    # creating a (reference cell numbers X query cell numbers) weights matrix for knn weights and unit weights
+    weights <- Matrix(0, nrow = ncol(object@raw.data[[reference]]), ncol = ncol(object@raw.data[[query]]), sparse = TRUE)
+    if (weight){ # for weighted situation
+      # find nearest neighbors for query cell in normed ref datasets
+      nn.k <- get.knnx(object@H.norm[reference_cells, ], object@H.norm[query_cells, ], k = knn_k, algorithm = "CR")
```

Comment:
This line should be pulled out before the `if` statement so that `nn.k` exists even if `weight = F`. Note that currently this causes the function to fail when weighting is set to FALSE. Probably worth testing a few different calls of this function in the unit tests so we can catch this more easily in the future!